google/cloud/apigee/registry/applications/v1/diff_analytics.proto

```diff
@@ -0,0 +1,88 @@
+syntax = "proto3";
+package google.cloud.apigee.registry.applications.v1;
+// [START go_declaration]
+option go_package = "github.com/apigee/registry/rpc;rpc";
+// [END go_declaration]
+
+/* What the Differ Module will compute, this proto contains
+the text diff of a spec and its revision */
+message TextDiff {
+  string diff = 1;
+  string spec = 2;
+  string spec_revision = 3;
+}
+
+/* What the Breaking Change Module will compute, this proto contains
+the diff separated into lines which represent individual changes.
+These changes are then labeled as a breaking change or non breaking change. */
+message DiffAnalysis {
+  //A single change
+  message ChangeLine{
+    enum ChangeType {
+      ADDED = 0;
+      MODIFIED = 1;
+      DELETED = 2;
+    }
+    /* A string line that represents 1 change
+   Ex : Components: schemas: modified: Pet: properties:
+   modified: name: type: from: string: to: "null" */
+    string line = 1;
+    //Added, Modified, Deleted
+    ChangeType change_type = 2;
+    bool breaking_change = 3;
+  }
+  repeated ChangeLine change_line = 1;
+  int64 change_count = 2;
+  int64 breaking_change_count = 3;
+}
+
+/* This represents the total stats of a resource.
+These stats are computed for all resources. */
+message Stats {
+  int64 changes = 1;
+  int64 breaking_changes = 2;
+  int64 revisions = 3;
+  int64 featurers_added = 4;
+  int64 features_removed = 5;
+}
+/* This represents computed metrics for a resource.
+These metrics are computed for all resources */
+message Metrics {
```

Comment:
This seems like it depends on the components and their APIs that don't exist yet. IIRC from the design, stats are computed in a different component (package?) of the system, then metrics are computed from the stats at a later stage. Something like `func ComputeStats(d DiffAnalysis) Stats` and `func ComputeMetrics(s Stats) Metrics` might exist.

If these were merged into a single message, would we go from DiffAnalysis directly to Stats+Metrics in one call? If that's the plan we might not need any of the stats fields at all since they're just intermediate values to derive the metrics.

Reply:
I could see wanting to keep these messages distinct so they could be attached as separate artifacts, which makes it possible to use the controller (with an appropriate manifest) to go from "Stats" to "Metrics", but if it's possible to compute this in one step, I think one step is probably better. 

Stepping back, I don't think "stats" and "metrics" will be well-understood descriptive terms for these chunks of data. Here "stats" seems to be the more basic (primitive) values, but "statistics" generally refers to analysis of populations of data. I think people would have trouble intuitively understanding that metrics are derived from stats and not the other way around. Bundling these together is one way to resolve that problem.